adding indexing / index advisor tutorial to user profile tutorial (#8)
couchbaselabs/tutorials-contrib: modules/session-storage/pages/udf.adoc

= Using User-Defined Functions with a Session Store

> :page-aliases: tutorials:session-storage-tutorial:udf

Reviewer: ```suggestion
```

Is that `:page-aliases:` needed? It's normally used to have a redirect, this one would suggest to redirect from this page https://docs.couchbase.com/tutorials/session-storage-tutorial/udf.html which doesn't exist.

Author: I have no idea why I put a page-alias in there. Cargo culting I guess? If it's safe to remove, I'll take it out in #9

== Overview

> This tutorial is part of the main Session Store tutorial (both link:http://docs.couchbase.com/tutorials/session-storage/aspnet.html[ASP.NET] and link:http://docs.couchbase.com/tutorials/session-storage/java.html[Java]). You should be able to follow along if you haven't gone through those tutorials yet, but reading those tutorials first will give you more context.

Reviewer: ```suggestion
This tutorial is part of the main Session Store tutorial (both xref:aspnet.adoc[ASP.NET] and xref:java.adoc[Java]). You should be able to follow along if you haven't gone through those tutorials yet, but reading those tutorials first will give you more context.
```

Author: Already fixed these in a different branch

  },
  {
    "dateCreated": "2019-08-01T14:22:17.4170499-04:00",
    "numItems": 1,
    "sessionStoreId": "47891218-fea5-ba04-34df-23c4e8dc70c6"
  },

  ...

]
----

`BASE64_DECODE` needs to be used in order to query individual properties of the encoded session data. For a complex query, this involves a lot of repetition.

A User-Defined Function (UDF) is can mitigate this repetition.

== Creating a User-Defined Function

To create user-defined functions, use `CREATE FUNCTION functionname() LANGUAGE INLINE AS ( ... );`.

The function can contain whatever N1QL you'd like. To better enable session storage queries, let's put the BASE64_DECODE inside of a function called `shoppingcarts`.

[source,SQL,indent=0]
----
CREATE FUNCTION shoppingcarts() LANGUAGE INLINE AS (
    SELECT VALUE OBJECT_PUT(BASE64_DECODE(s.shoppingcart), 'sessionStoreId', META(s).id)
    FROM sessionstore s
    WHERE s.shoppingcart IS NOT MISSING
);
----

The goal of this function is to return pre-decoded shopping cart data from the sessionstore bucket. Both `VALUE` and `OBJECT_PUT` are necessary in order to return a complete shopping cart document.

More information about these keywords:

> * link:https://docs.couchbase.com/server/6.0/n1ql/n1ql-language-reference/metafun.html[BASE64_DECODE] decodes the session store object

Reviewer: ```suggestion
* link:https://docs.couchbase.com/server/current/n1ql/n1ql-language-reference/metafun.html[BASE64_DECODE] decodes the session store object
```
Should it be `current`?
